File: 0138.复制带随机指针的链表/0138-复制带随机指针的链表.cpp
/**
 * Definition for singly-linked list with a random pointer.
 * struct RandomListNode {
 *     int label;
 *     RandomListNode *next, *random;
 *     RandomListNode(int x) : label(x), next(NULL), random(NULL) {}
 * };
 */
class Solution {
public:
    RandomListNode *copyRandomList(RandomListNode *head) {
        std::map<RandomListNode*, int> Node_map;
        std::vector<RandomListNode *> Node_vector;
        RandomListNode* ptr = head;
        int i = 0;
        while(ptr){
            Node_vector.push_back(new RandomListNode (ptr->label));
            Node_map[ptr] = i;
            i ++;
            ptr = ptr->next;
        }
        Node_vector.push_back(0);
        ptr = head;
        i = 0;
        while(ptr){
            Node_vector[i]->next = Node_vector[i+1];
            if(ptr->random){
                Node_vector[i]->random = Node_vector[Node_map[ptr->random]];
            }
            i++;
            ptr = ptr->next;
        }
        return Node_vector[0];
    }
};

File: 0138.复制带随机指针的链表/0138-复制带随机指针的链表.py
"""
# Definition for a Node.
class Node(object):
    def __init__(self, val, next, random):
        self.val = val
        self.next = next
        self.random = random
"""
class Solution(object):
    def copyRandomList(self, head):
        """
        :type head: Node
        :rtype: Node
        """
        #133变种题， 图换成链表
        mapping = dict()
        
        p = head
        while p:
            mapping[p] = Node(p.val, None, None)
            p = p.next
            
        for key, val in mapping.items(): #key是老结点， val是新节点
            if key.next:
                val.next = mapping[key.next]
            if key.random and key.random in mapping:
                val.random = mapping[key.random]
            
        return mapping[head] if head else head
        
        